### ISSUE TITLE
Allow IAuthtenticator methods to modify the response

### ISSUE THREAD

**Issue by amercader** [ckan/ckan#5259]
The Pylons standard way of modifying the outgoing response (eg adding cookies, issuing redirects etc) is to either edit a global `response` object or call global functions like `redirect_to`.

Flask views on the contrary always expect a response-like object to be returned so if you want to modify it, it needs to be in the context of that view function. Take for instance the `login` function of the `user` view:

```python
  def logout():
      # Do any plugin logout stuff
      for item in plugins.PluginImplementations(plugins.IAuthenticator):
          item.logout()
      url = h.url_for(u'user.logged_out_page')
      return h.redirect_to(
          _get_repoze_handler(u'logout_handler_path') + u'?came_from=' + url,
          parse_url=True)
```
It's quite common for `IAuthenticator.login()` or `IAuthenticator.logout()` to need to clear cookies, issue redirects etc. In Pylons, plugins could just do `toolkit.response.cookies = ...` or `toolkit.redirect_to()` and the web response would be immediately affected.

On Flask though the current code just calls the plugin methods and carries on with the normal view function execution, so extensions can't do any of these actions.

One potential way I've solved it for this particular case is checking if extensions are returning a response object, and if so just returning this directly on the view function, as it assumes that the extension has done all that's needed.

Happy to hear alternatives.

**Comment by pazepaze** [comment]
I am currently working on a fork of [ckanext-oauth2 plugin](https://github.com/conwetlab/ckanext-oauth2) and would appreciate a solution for this.

I need to be able to set cookies when IAuthenticator.identify() is called from flask.
I also need to trigger a logout and my current solution for this would be to redirect to /logout, which I can't do from flask.

**Comment by pazepaze** [comment]
I made a draft pull request that allows redirects from ```IAuthenticator.identify()```. I did not touch ```IAuthenticator.login()``` or ```IAuthenticator.logout()```. I also can't set cookies yet.

Can we set cookies to some sort of request context? Maybe ```toolkit.g```? I tried setting values there but could not figure out how to write to this object. I also don't really know where I would put the code that retrieves those cookies from the context and puts them into the response at the end.
